### PR TITLE
Implement `String()` for the state so it can be easily printed.

### PR DESCRIPTION
**Pull request by kalbasit**

### CODE REVIEW

File: gobreaker.go [hunk 1]
```diff
@@ -9,15 +9,29 @@ import (
...
 const (
-	StateClosed   State = iota
-	StateHalfOpen State = iota
-	StateOpen     State = iota
+	StateClosed State = iota
+	StateHalfOpen
+	StateOpen
 )
 
+// String implements stringer interface
```
Comment (shurcooL) [inline gobreaker.go]: Minor, but this sentence is missing a period at the end. Which is inconsistent with all other comments here.

**Comment by shurcooL** [comment]
All other changes LGTM. I'm not very familiar with this project though.

File: gobreaker.go (hunk 1 above)
Comment (kalbasit) [inline gobreaker.go]: done

File: gobreaker.go [hunk 2]
```diff
@@ -9,15 +9,29 @@ import (
 )
 
 // State is a type that represents a state of CircuitBreaker.
-type State int
+type State uint8
```
Comment (kalbasit) [inline gobreaker.go]: I'll revert this change and file it in a separate PR.

Comment (kalbasit) [inline gobreaker.go]: done

**Comment by YoshiyukiMineo** [comment]
Hi

We'd like to avoid panic as possible:
panic("unknown state") -> return fmt.Sprintf("unknown state: %d", s)
See [Effective Go](https://golang.org/doc/effective_go.html#panic).

Could you squash your commits into one commit?

**Comment by kalbasit** [comment]
Hi,

So the state values are 

```
 const (
	StateClosed State = iota
	StateHalfOpen
	StateOpen
 )
```

anything else is not a "possible value". So `panic` is pretty much is the only way to avoid a return keyword at the end, like so:

```
func (s State) String() string {
	switch s {
	case StateClosed:
		return "closed"
	case StateHalfOpen:
		return "half-open"
	case StateOpen:
		return "open"
	default:
		panic("unknown state")
	}
	
	return ""
}
```

which is invalid because "the state of impossible is not empty string". The only way to signal this impossible value (which should only come from memory corruption or a compiler bug) is to `panic`.

**Comment by shurcooL** [comment]
> The only way to be assigned such value would be a compiler bug or a faulty memory.

That's not true. Another way is to write faulty but correct (as in, it will be compiled without errors) code:

```Go
var foo = State(7) // Not a known/supported value, but it's valid Go code.
foo.String()
```

**Comment by kalbasit** [comment]
@shurcooL, that would be an "impossible value" which should not go unnoticed. Please see [this example](https://github.com/golang/go/blob/782db7fc883e3ae409b27f7b77b135d8eee81c69/src/image/jpeg/scan.go#L37-L39) from the Go source code. It is true that assigning an unknown state is "valid Go code" but not valid in the context of the this state machine.

I would argue otherwise if the [`fmt.Stringer`](https://godoc.org/fmt#Stringer) returns an error,

**Comment by shurcooL** [comment]
There's one distinction to be made.

Had the type `State` been unexported, then its value could be ensured to never contain an invalid value by the internal code of the package.

However, `gobreaker.State` is an exported identifier that the _user_ of the package could use incorrectly, and when printing, it will panic.

All the examples you've shown, as far as I can tell, only apply to internal enums, so unless the package authors make a mistake, the user can never get a `panic`. Is that right?

On the other hand, because `gobreaker.State` is exported, the package cannot ensure the user won't get a panic if they use it incorrectly.

Just pointing out an observed difference.

**Comment by kalbasit** [comment]
You'll probably right on the fact that it is exported. It is however a problem because it's a state machine that will be invalid if the state was set to a value that is not expected.

At this point we could:
- leave the panic as is. It is important to tell the user that the state machine is an invalid state.
- replace the panic with `return fmt.Sprintf("unexpected value %d", int(s))`

@YoshiyukiMineo which particular path you'd like?

**Comment by YoshiyukiMineo** [comment]
The state machine never falls into a invalid state because the users can only get the copy of the state and cannot set the state of the machine.

State.String is a simple method that just converts a state to a string and runs outside the state machine.
So I think it should not panic and I'd like to replace it with `return fmt.Sprintf("unknown state: %d", s)`.

**Comment by kalbasit** [comment]
done.

**Comment by YoshiyukiMineo** [comment]
Thank you.

Could you create a new pull request excluding the unrelated commit, "gobreaker: uint8 is enough for the State type"?

**Comment by kalbasit** [comment]
@YoshiyukiMineo no need to create a separate PR, I squashed the commits into one. PTAL.

**Comment by YoshiyukiMineo** [comment]
Thank you very much.

**Pull request merged**